src/rdkafka_int.h

```diff
@@ -912,7 +918,46 @@ void rd_kafka_log0(const rd_kafka_conf_t *conf,
                 }                                              \
         } while (0)
 
-
+/*
+        Wait for available memory space before allocating the expected memory size.
+        We use compare-and-swap (CAS) to register the memory one thread wants to allocate. CAS only
+        succeed if the total memory usage is still the same when checking. If another thread
+        already register its expected memory usage, the current thread needs check for the total
+        memory usage again.
+*/
+#define rd_wait_mem_space(expected_size, succeeded)  do {                                               \
+                int64_t current_total_mem_usage;                                                        \
+                rd_bool_t counted = rd_false;                                                           \
+                int32_t wait_time = 0;                                                                  \
+                succeeded = rd_true;                                                                    \
+                do {                                                                                    \
+                        /* if the total memory usage is too high, we need to wait */                    \
+                        current_total_mem_usage = rd_atomic64_get(&rk_total_message_mem);               \
+                        while(unlikely(current_total_mem_usage + (expected_size) >                      \
+                                rd_atomic64_get(&rk_total_message_mem_limit))) {                        \
+                                if (counted == rd_false) {                                              \
+                                        counted = rd_true;                                              \
+                                        rd_atomic32_add(&rd_num_wait_space_threads, 1);                 \
+                                }                                                                       \
+                                if (unlikely(wait_time >= 10)) {                                        \
```

Comment:
Should constants related to wait_time be configured?

Reply:
I can add a variable for this. But normally, we do not change this one very often